mixin-metrics.go

```diff
@@ -0,0 +1,319 @@
+
+	return nil
+}
+
+// use jq on a file to extract prom queries
+func ParseJq(queries *[]string, jsonData map[string]interface{}, jqExpr string) error {
+
+	query, err := gojq.Parse(jqExpr)
+	if err != nil {
+		return err
+	}
+
+	iter := query.Run(jsonData)
+	for {
+		v, ok := iter.Next()
+		if !ok {
+			break
+		}
+		if err, ok := v.(error); ok {
+			return err
+		}
+		*queries = append(*queries, v.(string))
+	}
+
+	return nil
+}
+
+// parses through a dashboard to extract queries, and then metrics from queries
+func ParseDashboard(file *os.File) (*MetricsFile, error) {
+
+	queries := make([]string, 0)
+	metrics := map[string]struct{}{}
+	errors := make([]error, 0)
+	res := map[string]interface{}{}
+
+	bytes, err := ioutil.ReadAll(file)
+	if err != nil {
+		return nil, err
+	}
+
+	json.Unmarshal([]byte(bytes), &res)
+	if err := ParseJq(&queries, res, ".templating.list[].query"); err != nil {
+		return nil, err
+	}
+
+	if err := ParseJq(&queries, res, ".panels[]?.targets[]?.expr"); err != nil {
+		return nil, err
+	}
+
+	if err := ParseJq(&queries, res, ".rows[].panels[].targets[].expr"); err != nil {
+		return nil, err
+	}
+
+	for _, query := range queries {
+		if err := ParseQuery(query, metrics); err != nil {
+			errors = append(errors, err)
+		}
+	}
+
+	metricsFile := NewMetricsFile(file.Name(), metrics, errors)
+	return &metricsFile, nil
+}
+
+// todo: separate rules and raw metrics
+// parses through a rules file and extracts queries, and then metrics from queries
+func ParseRules(file *os.File) (*MetricsFile, error) {
+
+	metrics := map[string]struct{}{}
+	errors := make([]error, 0)
+
+	var conf RuleConfig
+
+	rulesFile, err := ioutil.ReadAll(file)
+	if err != nil {
+		return nil, err
+	}
+
+	err = yaml.Unmarshal(rulesFile, &conf)
+	if err != nil {
+		return nil, err
+	}
+
+	groups := conf.RuleGroups
+	for _, group := range groups {
+		for _, rule := range group.Rules {
+			err := ParseQuery(rule.Expr, metrics)
+			if err != nil {
+				errors = append(errors, err)
+			}
+		}
+	}
+
+	metricsFile := NewMetricsFile(file.Name(), metrics, errors)
+
+	return &metricsFile, nil
+}
+
+// iterates over all rules/dash files in dir
+func ParseDir(dir string, isRules bool) (*MetricsDir, error) {
+
+	files, err := ioutil.ReadDir(dir)
+	if err != nil {
+		return nil, err
+	}
+
+	metricsDir := MetricsDir{
+		MetricsFiles: make([]MetricsFile, 0),
+	}
+
+	for _, fileInfo := range files {
+		fmt.Println("Parsing: ", fileInfo.Name())
+		file, err := os.Open(dir + "/" + fileInfo.Name())
+		if err != nil {
+			return nil, err
+		}
+
+		if isRules {
+			metricsFile, err := ParseRules(file)
+			if err != nil {
+				return nil, err
+			}
+			metricsDir.MetricsFiles = append(metricsDir.MetricsFiles, *metricsFile)
+		} else {
+			metricsFile, err := ParseDashboard(file)
+			if err != nil {
+				return nil, err
+			}
+			metricsDir.MetricsFiles = append(metricsDir.MetricsFiles, *metricsFile)
+		}
+
+		err = file.Close()
+		if err != nil {
+			return nil, err
+		}
```

Comment:
since you are returning in multiple places (L236, L242), usually you want to use a defer for example right after line 231.

```
defer file.Close()
```

this doesn't return an error if Close returns an error. If that's important to your case, then you need to make sure that you are calling file.Close() in all the places where you return, or use a feature (which is somewhat frowned upon) called "named returns".